### ISSUE TITLE
Move memoization into a separate library

### ISSUE THREAD

**Issue by sferik** [dkubb/adamantium#20]
I recently [added `adamantium`](https://github.com/sferik/twitter/commit/c735bbccc3ff82f3a4e31c6ab09be4140f33d2b9) to my project but had to [revert](https://github.com/sferik/twitter/commit/1a06126c66102817f1346b69ceccf8bd2a564d4b) that change, since this library (specifically, the freezing behavior) isn't compatible with Ruby 1.8.7, which I need to continue to support.

However, I the memoization code is fully 1.8.7 compatible and I was able to [copy it into my project with just a few modifications](https://github.com/sferik/twitter/commit/1a06126c66102817f1346b69ceccf8bd2a564d4b#diff-7). It would be nice if this was a separate gem for people who want to use it.

Alternatively, consider supporting Ruby 1.8.7 (but that's probably more work and less desirable in the long term).

**Comment by sferik** [comment]
Upon further inspection, there is already a [`memoizable` Ruby library](https://github.com/ecomba/memoizable). Any reason you're not using this for memoization in `adamantium`?

**Comment by sferik** [comment]
I may have spoken too soon… :speak_no_evil:

I just looked [the code for `memoizable`](https://github.com/ecomba/memoizable/blob/master/lib/memoizable.rb) (it's just that one file) and it's not pretty (nor is it thread safe). [This one](https://github.com/djberg96/memoize/blob/master/lib/memoize.rb) looks about the same.

I still think there's a need for a modern memoizable Ruby library that's thread safe and modular enough to integrate with `adamantium`. What do you think about submitting a pull request to one of these existing projects? I could work on this, if you're interested.

**Comment by mbj** [comment]
@sferik 

1.8 - support:

Unlikely we can readd support for 1.8. Our tooling contains lots of tools that are > 1.8 only. Also 1.8 is EOL, maybe you can drop 1.8 support also?

3rd party memoizers:

I think we need our own. The memoizer does have to work on frozen objects, and it must only provide memoization for query methods with zero arguments. I think all available memoizers do to much or simply do not work in our environment.

extraction:

Yeah. I'd support extraction of memoizer. Ideally you could use our `devtools` metagem and the same metrics system. Good names are squatted already. But our memoization is more about idempotency and lazyness. So `idempotent` could be a gem name.

**Comment by dkubb** [comment]
I would support extracting memoization from adamantium. It all started off as something really simple that I used internally, but grew to the point where I can see it being useful in a context outside of adamantium. I've used a few of the other memoization libraries, like memoist, and while I liked the functionality the implementations are usually pretty crazy.

@sferik If you're interested I could register the name `idempotent`, and make a repo for the work of extraction to begin in.

As for supporting 1.8, I certainly wouldn't do anything to actively block making it 1.8 compatible, but if there were ever conflicts between 1.8 and 1.9/2.0, I would likely go in favour of the latter. In most of our gems (ROM related and personal ones) we dropped 1.8 explicit support when it went EOL, and I was probably the last hold-out in the team so I can understand where you're coming from. What's your specific use case for wanting to support 1.8?

**Comment by sferik** [comment]
@dkubb Please do register the name `idempotent` and add me to the repo. 

My specific reason for needing to support Ruby 1.8.7 is that—as far as I know—Twitter itself is still running on [Kiji](https://github.com/twitter-forks/rubyenterpriseedition187-248/) a fork of Ruby Enterprise Edition based on MRI 1.8.7-p248, with improved garbage collection (similar to the sunny/shady generational GC coming in Ruby 2.1.0). The `twitter` gem is used internally at Twitter Inc and I’d like them to be able to continue using it, as long as they're on Kiji.

I actually [recently removed support for Ruby 1.8.7](https://github.com/sferik/twitter/commit/0ed8d32c7c237be3ab5d7874c4be1df1f925d387) but ended up [reverting](https://github.com/sferik/twitter/commit/120648ee7dcadb6982c0b041edbb9811a85749d0) that change. Since it’s not so hard to maintain, there’s no compelling reason to stop supporting it, at least through the next major version of the gem. In addition to Twitter itself, there are a significant number of legacy Ruby applications still running on Ruby 1.8.7, as confirmed by the rubygems.org server logs.

**Comment by dkubb** [comment]
> Please do register the name idempotent and add me to the repo.

Ok, I will do that tonight after work.

I understand the use case for 1.8.7 support. I saw your tweets asking if Twitter itself still runs 1.8, hopefully we'll find out from someone internally what their state is.

Like I said, I don't mind supporting 1.8.7 as long as there's a valid use case and it's clear to everyone that if maintaining compatibility with 1.8.7 ever blocks the code from working in 1.9/2.0+ that support would be dropped. I don't foresee anything like that in this case, given there are almost always work-arounds, and this gem will be pretty simple, but it's good to be explicit about it up-front.

I have one more question about memoization, and whether or not your usage would need to support memoizing methods with a nonzero arity? We discussed this today in another PR: https://github.com/dkubb/adamantium/pull/21

**Comment by sferik** [comment]
Shortly after posting this comment, I came across this [article in Wired](http://www.wired.com/wiredenterprise/2013/09/the-second-coming-of-java/all/), implying that Twitter has all but completely moved away from Ruby. This is [confirmed by @mitchellh](https://twitter.com/mitchellh/status/382941373743501312), who tends to know stuff.

So maybe supporting Ruby 1.8.7 isn't that important but I see no reason not to.

As for memoizing methods with non-zero arity, this is not necessary for my use-case but I could imagine supporting it in future versions of the `idempotent` gem.

**Comment by misfo** [comment]
Memoizing methods on immutable objects is a lot simpler than memoizing methods on mutable objects. No cache invalidation to deal with! And we know what a rat's nest that can be:

```
There are only two hard things in Computer Science: cache invalidation and naming things.

-- Phil Karlton
```

Some of the simplicity of the memoization implementation in this repo can be credited to the fact that the objects are guaranteed not to change. Would the extracted library have to support mutable and immutable objects?

**Comment by dkubb** [comment]
@sferik ok, idempotent was taken but I released idem 0.0.0 at rubygems for now.

I've created an initial WIP branch for experimenting with the interface and getting the initial layout right here: https://github.com/dkubb/idem/pull/1

I'm going to mark this issue as resolved and move the discussion over into that PR.
